ext/reflection/php_reflection.c

```
@@ -7321,26 +7321,59 @@ ZEND_METHOD(ReflectionAttribute, newInstance)
+		attr->attributes,
+		"delayedtargetvalidation",
+		strlen("delayedtargetvalidation")
+	);
 
-		if (!(attr->target & flags)) {
-			zend_string *location = zend_get_attribute_target_names(attr->target);
-			zend_string *allowed = zend_get_attribute_target_names(flags);
+	/* This code can be reached under one of three possible conditions:
+	 * - the attribute is an internal attribute, and it had the target and
+	 *   and repetition validated already
+	 * - the attribute is an internal attribute and repetition was validated
+	 *   already, the internal validator might have been run if the target was
+	 *   correct, but any error would have been stored in
+	 *   `zend_attribute.validation_error` instead of being thrown due to the
+	 *   presence of #[DelayedTargetValidation]
+	 * - the attribute is a user attribute, and neither target nor repetition
+	 *   have been validated.
+	 */
+	uint32_t flags = zend_attribute_attribute_get_flags(marker, ce);
+	if (EG(exception)) {
+		RETURN_THROWS();
+	}
 
-			zend_throw_error(NULL, "Attribute \"%s\" cannot target %s (allowed targets: %s)",
-				ZSTR_VAL(attr->data->name), ZSTR_VAL(location), ZSTR_VAL(allowed)
-			);
+	/* No harm in always running target validation, for internal attributes
+	 * without #[DelayedTargetValidation] it isn't necessary but will always
+	 * succeed. */
+	if (!(attr->target & flags)) {
+		zend_string *location = zend_get_attribute_target_names(attr->target);
+		zend_string *allowed = zend_get_attribute_target_names(flags);
 
-			zend_string_release(location);
-			zend_string_release(allowed);
+		zend_throw_error(NULL, "Attribute \"%s\" cannot target %s (allowed targets: %s)",
+			ZSTR_VAL(attr->data->name), ZSTR_VAL(location), ZSTR_VAL(allowed)
+		);
+
+		zend_string_release(location);
+		zend_string_release(allowed);
+
+		RETURN_THROWS();
+	}
 
+	/* Report the delayed validator error for internal attributes */
+	if (delayed_target_validation && ce->type == ZEND_INTERNAL_CLASS) {
```

Don't think we need to load this at all. When `validation_error` is set, we can just throw the error.

---

Similar thinking as above. It's true that `validation_error` is currently only set for internal classes and with `#[DelayedTargetValidation]`, but verifying this fact is not necessary for this code to be correct. More code generally is more cognitive overhead.

Regardless, if you think this check is useful, I think it makes more sense to invert it:

* Check whether `validation_error` is set.
* Then, in `#if ZEND_DEBUG`, load `delayed_target_validation` and assert that `delayed_target_validation && ce->type == ZEND_INTERNAL_CLASS`.
* And then emit the error.

---

makes sense, done